### PR TITLE
Added send mail functionality to send token via mail

### PR DESCRIPTION
**Pull request by fanyui**
## Description of what I changed

Add possibility for sending email with activation key 
## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-

## Checklist: I completed these to help reviewers :)

- [ x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [ x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by fanyui** [comment]
@wluyima  This pull request sends the emai with generated activation key. so we can use it to test the rest webservice endpoint.

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 1]
```diff
@@ -764,10 +771,59 @@ public User setUserActivationKey(User user) {
 		credentials.setActivationKey(activationKey);	
 		dao.setUserActivationKey(credentials);	
 			
-		//Send Email with unhashed  activation key and Date:
-		
+		//Send Email with unhashed  activation key
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: I think you can get rid of this comment

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -764,10 +771,59 @@ public User setUserActivationKey(User user) {
...
 		dao.setUserActivationKey(credentials);	
 			
-		//Send Email with unhashed  activation key and Date:
-		
+		//Send Email with unhashed  activation key 
+		sendMail(user, token);
 		return user;
 	}
+	
+	public void sendMail(User user, String token) {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: You don't need this method and all this code, I did mention there is a messaging service and you should use it to send the email.

Comment (wluyima) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: This should be a one line call

Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Hello @wluyima  I've been looking at the MessageService to use  that service method and send the email but. when i tried it fails with errors of **Must issue a STARTTLS command first. v10-v6sm19344157wrm.18 - gsmtp** and i guess the MessageService is for only notification and alerts. as when i looked at the MailMessageSender. it has methods setMailSession which takes a session variable and creating a session requires the Propertites and an authenticator which is what i did in the code above. I was hopping to see some where. how they use the Global variables defined in the OpenmrsConstants. but i couldn't find any.
Thanks for any directions.

Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: I did use Context.getMessageService().sendMessage(mail) where mail is a message object ; and since no session is set message always fails

Comment (wluyima) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: There is a sendMessage method that doesn't take a message object, I would try that first.

Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: The failure is when am still building so by then am not yet logged in. i will do some digging again and see.

**Comment by wluyima** [comment]
Let's not do localization for now, it will be done in the future

File: api/src/main/java/org/openmrs/api/context/Context.java [hunk 3]
```diff
@@ -567,6 +567,7 @@ private static Session getMailSession() {
 					props.setProperty("mail.from", adminService.getGlobalProperty("mail.from"));
 					props.setProperty("mail.debug", adminService.getGlobalProperty("mail.debug"));
 					props.setProperty("mail.smtp.auth", adminService.getGlobalProperty("mail.smtp_auth"));
+					props.setProperty(OpenmrsConstants.GP_MAIL_SMTP_STARTTLS_ENABLE, adminService.getGlobalProperty(OpenmrsConstants.GP_MAIL_SMTP_STARTTLS_ENABLE));
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/context/Context.java]: I said that you need to create a separate TRUNK ticket to address because it's a bug in core, then create a separate PR to openmrs:master instead to get it in, this change should not be in the branch for your GSoC project

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -763,9 +765,26 @@ public User setUserActivationKey(User user) {
 		LoginCredential credentials = dao.getLoginCredential(user);
 		credentials.setActivationKey(activationKey);	
 		dao.setUserActivationKey(credentials);	
-			
-		//Send Email with unhashed  activation key and Date:
 		
+		String link = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GP_HOST_URL) + token;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: this is not correct, the URL should have a placeholder to replace the token, you should not be appending it

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -763,9 +765,26 @@ public User setUserActivationKey(User user) {
...
+		String msg = "Hi " + user.getUsername()
+		        + ", \n It looks like you like to change your password, click on the following link to do so.\n" + link
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Somewhat a lot of repetition of **like**, can you rephrase this please?

Comment (wluyima) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Do line breaks really work here? I would think html would

Comment (wluyima) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: For a start, add a file to the classpath containing this message and read in the contents when sending the email, in the feature it will be a template that can support localization too

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -763,9 +765,26 @@ public User setUserActivationKey(User user) {
...
+		        + " please disregard this email if you didn't innitiate the password reset process  The openmrs Team";
+		Message mail = new Message();
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: You should not be constructing anything, please use the message service, all this should be like 2 calls to the service once you have the contents of the email.

File: api/src/main/java/org/openmrs/api/context/Context.java (hunk 3 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/context/Context.java]: Ok i did it already this was just to permit me do my testing and ensure it works with respect to my gsoc project. pull request for the issue is https://github.com/openmrs/openmrs-core/pull/2703

Comment (wluyima) [reply api/src/main/java/org/openmrs/api/context/Context.java]: Then remove it please

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 7]
```diff
@@ -1390,12 +1392,17 @@ public void getUserByUsernameOrEmail_shouldFailIfEmailIsNull() {
...
 	
 	@Test
-	public void setUserActivationKey_shouldCreateUserActivationKey() {
+	public void setUserActivationKey_shouldCreateUserActivationKey() throws MessageException {
 		User u = new User();
 		u.setPerson(new Person());
 		u.addName(new PersonName("Benjamin", "A", "Wolfe"));
 		u.setUsername("bwolfe");
 		u.getPerson().setGender("M");
+		Context.getAdministrationService().setGlobalProperty("mail.smtp_auth", "false");
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: Why is this necessary

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 8]
```diff
@@ -1390,12 +1392,17 @@ public void getUserByUsernameOrEmail_shouldFailIfEmailIsNull() {
...
+		Context.getAdministrationService().setGlobalProperty(OpenmrsConstants.GP_MAIL_SMTP_STARTTLS_ENABLE, "true");
+		Context.getAdministrationService().setGlobalProperty("mail.smtp_host", "localhost");
+		Context.getAdministrationService().setGlobalProperty("mail.smtp_port", "25");
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: Why are these GP necessary?

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 9]
```diff
@@ -689,6 +694,10 @@ public static String getOpenmrsProperty(String property) {
 		props.add(new GlobalProperty(GP_MAIL_SMTP_STARTTLS_ENABLE, "false",
 		        "Set to true to enable TLS encryption, else set to false"));
 		
+		props.add(new GlobalProperty(GP_HOST_URL,
+		        "http://localhost:8080/openmrs/admin/users/changePassword.form?activationkey=",
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: You can remove this default value for now, it will be added once we update the form

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -691,9 +713,91 @@ public String getSecretQuestion(User user) throws APIException {
...
+	 * @see org.openmrs.api.UserService#getUserByActivationKey(java.lang.String)
+	 * 
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public User getUserByActivationKey(String activationKey) {
+		LoginCredential loginCred = dao.getLoginCredentialByActivationKey(activationKey);
+		if (loginCred != null) {
+			String[] credTokens = loginCred.getActivationKey().split(":");
+			if (System.currentTimeMillis() <= Long.parseLong(credTokens[1])) {
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: What happens if the credTokens length is 0?

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -691,9 +713,91 @@ public String getSecretQuestion(User user) throws APIException {
...
+	/**
+	 * @see org.openmrs.api.UserService#getUserByActivationKey(java.lang.String)
+	 * 
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public User getUserByActivationKey(String activationKey) {
+		LoginCredential loginCred = dao.getLoginCredentialByActivationKey(activationKey);
+		if (loginCred != null) {
+			String[] credTokens = loginCred.getActivationKey().split(":");
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: what happens if the `getActivationKey()` return null for an instance?

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -691,9 +713,91 @@ public String getSecretQuestion(User user) throws APIException {
...
+		mail.setContent(msg);
+		mail.setSentDate(new Date());
+		
+		try {
+			Context.getMessageService().sendMessage(mail);
+		}
+		catch (MessageException e) {
+			// Prevent propagation of exception to the webservice should in case message was not sent  
+		}
+		return user;
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: So up to now, How the method will indicate that the email was sent without any faults to the caller? Do we want to indicate the caller about the email success or not?

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -691,9 +713,91 @@ public String getSecretQuestion(User user) throws APIException {
...
+		credentials.setActivationKey(activationKey);	
+		dao.setUserActivationKey(credentials);	
+		
+		String link = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GP_HOST_URL) + token;
+		String msg = "Hi " + user.getUsername()
+		        + ", \n It looks like you like to change your password, click on the following link to do so.\n" + link
+		        + " please disregard this email if you didn't innitiate the password reset process  The openmrs Team";
+		Message mail = new Message();
+		mail.setId(20);
+		mail.setRecipients(user.getEmail());
```
[the remark this answers is not in the document]
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: This method expects a user and it the responsibility of the caller to provide the user. Should incase null is provided, the mail will fails with message exception which in either case we don't want to inform the user of the failure. so everything goes on well

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 14]
```diff
@@ -691,9 +713,91 @@ public String getSecretQuestion(User user) throws APIException {
...
+		String token = RandomStringUtils.randomAlphanumeric(20);
+		long time = System.currentTimeMillis() + getValidTime();
+		String hashedKey = Security.encodeString(token);
+		String activationKey = hashedKey+":"+time;
+		LoginCredential credentials = dao.getLoginCredential(user);
+		credentials.setActivationKey(activationKey);	
+		dao.setUserActivationKey(credentials);	
+		
+		String link = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GP_HOST_URL) + token;
+		String msg = "Hi " + user.getUsername()
```
[the remark this answers is not in the document]
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Ok, Am working on this.

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 12 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Users needs not know of the success or failure of the mails

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 15]
```diff
@@ -691,9 +713,91 @@ public String getSecretQuestion(User user) throws APIException {
...
+	/**
+	 * @throws MessageException
+	 * @see org.openmrs.api.UserService#setUserActivationKey(org.openmrs.User)
+	 */
+	@Override
+	public User setUserActivationKey(User user) {
+		String token = RandomStringUtils.randomAlphanumeric(20);
+		long time = System.currentTimeMillis() + getValidTime();
+		String hashedKey = Security.encodeString(token);
+		String activationKey = hashedKey+":"+time;
```
[the remark this answers is not in the document]
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Ok

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 11 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Trusting the code that generation for the activation key will always be provided ie  setUserActivationKey(User user)  as that is it job.  and for this line of code to be executed the loginCred is not null which was gotten using the activation key hence we are guaranteed that if loginCred exist and is not null then activation key exist else null will be returned

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 10 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: credTokens length cannot be null at this level because if loginCred is not null then there is an activation key of the form xddkkdud7i8d:55636737352 which ensures that credToken lenght is always greater than 0, other wise  null will be returned instead of a failure

File: api/src/main/java/org/openmrs/api/context/Context.java (hunk 3 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/context/Context.java]: Ok

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 4 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Ok

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 5 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: @wluyima Line breaks work instead of html <br /> tags. Please am having issues adding txt files to classpath.

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 16]
```diff
@@ -751,21 +753,34 @@ public User getUserByActivationKey(String activationKey) {
...
-		String activationKey = hashedKey+":"+time;
+		String activationKey = hashedKey + ":" + time;
 		LoginCredential credentials = dao.getLoginCredential(user);
 		credentials.setActivationKey(activationKey);	
 		dao.setUserActivationKey(credentials);	
-			
-		//Send Email with unhashed  activation key and Date:
 		
+		String link = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GP_HOST_URL)
+		        .replace("activationkey", "activationkey=" + token);
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Please try to use constants where necessary, the placeholder needs to look explicit, use something like  **{actvitationKey}**, and you should replace the placeholder with only the key, nothing else.

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 17]
```diff
@@ -751,21 +753,34 @@ public User getUserByActivationKey(String activationKey) {
...
+		        + link
+		        + "\n Please disregard this email if you didn't innitiate the password reset process.  \n The openmrs Team";
+		
+		try {
+			Message mail = Context.getMessageService().createMessage(user.getEmail(),
+			    Context.getAdministrationService().getGlobalProperty("mail.from"), "Password Reset Request", msg);
+			Context.getMessageService().sendMessage(mail);
+		}
+		catch (MessageException e) {
+			// Prevent propagation of exception to the webservice should in case message was not sent
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: You shouldn't be swallowing the exception

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 18]
```diff
@@ -751,21 +753,34 @@ public User getUserByActivationKey(String activationKey) {
...
+		        .replace("activationkey", "activationkey=" + token);
+		String msg = "Hi " + user.getUsername()
+		        + ",\n It looks like you want to change your password, click on the following link to do so.\n"
+		        + link
+		        + "\n Please disregard this email if you didn't innitiate the password reset process.  \n The openmrs Team";
+		
+		try {
+			Message mail = Context.getMessageService().createMessage(user.getEmail(),
+			    Context.getAdministrationService().getGlobalProperty("mail.from"), "Password Reset Request", msg);
+			Context.getMessageService().sendMessage(mail);
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: There is actually a sendMessage method that takes in recipients, from, subject and content, I think you need to call that.

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 17 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: In the restwebservice A post for request of passwordreset  will trigger this method and that post is supposed to always return http.OK  and if the exception is not handled here it will propergate back as the response and we end up not having an ok response.

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (suthagar23) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: I think, you will get NPE in L775?

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 12 above)
Comment (suthagar23) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: So if someone did not get the mail, then how can the user or System Administrator will get to know about that?
Is it good to have a debug log about this?

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 19]
```diff
@@ -689,6 +694,9 @@ public static String getOpenmrsProperty(String property) {
 		props.add(new GlobalProperty(GP_MAIL_SMTP_STARTTLS_ENABLE, "false",
 		        "Set to true to enable TLS encryption, else set to false"));
 		
+		props.add(new GlobalProperty(GP_HOST_URL, "",
+		        "The activationKey is the parameter which will be generated and substituted. always make sure to provide a parameter name activationkey in the link you provide here "));
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: A full stop can't be followed by a lower case letter

Comment (wluyima) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: This is actually wrong, don't dictate that the URL should have a parameter because that's a implementation detail

Comment (wluyima) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: You should provide a proper description of the GP, this doesn't feel like one

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 17 above)
Comment (wluyima) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: You should not swallow the exception here, that is implementation logic in REST and should have nothing to do with the service layer

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 20]
```diff
@@ -766,20 +768,22 @@ public User setUserActivationKey(User user) {
...
 		dao.setUserActivationKey(credentials);	
 		
-		String link = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GP_HOST_URL)
-		        .replace("activationkey", "activationkey=" + token);
-		String msg = "Hi " + user.getUsername()
-		        + ",\n It looks like you want to change your password, click on the following link to do so.\n"
-		        + link
-		        + "\n Please disregard this email if you didn't innitiate the password reset process.  \n The openmrs Team";
+		Properties props = new Properties();
+		InputStream stream = OpenmrsClassLoader.getInstance().getResourceAsStream("emailTemplate.properties");
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: The file isn't supposed to be a properties file, otherwise you might as well move everything to the messages.properties file which already supports internationalization anyways and get rid of this file.

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 21]
```diff
@@ -751,21 +755,36 @@ public User getUserByActivationKey(String activationKey) {
...
+		
+		String link = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GP_HOST_URL)
+		        .replace("{activationKey}", token);
+		user.getUsername();
+		String msg = props.getProperty("mail.content").replaceAll(":name", user.getUsername()) + link;
+		try {
+			Context.getMessageService().sendMessage(user.getEmail(),
+			    Context.getAdministrationService().getGlobalProperty("mail.from"), props.getProperty("mail.subject"), msg);
+		}
+		catch (MessageException e) {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Please don't catch he exception

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 22]
```diff
@@ -751,21 +755,36 @@ public User getUserByActivationKey(String activationKey) {
...
 		String hashedKey = Security.encodeString(token);
-		String activationKey = hashedKey+":"+time;
+		String activationKey = hashedKey + ":" + time;
 		LoginCredential credentials = dao.getLoginCredential(user);
 		credentials.setActivationKey(activationKey);	
 		dao.setUserActivationKey(credentials);	
-			
-		//Send Email with unhashed  activation key and Date:
 		
+		Properties props = new Properties();
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: The file isn't supposed to be a properties file, otherwise you might as well move everything to the messages.properties file which already supports internationalization anyways and get rid of this file.

File: api/src/main/resources/emailTemplate.properties [hunk 23]
```diff
@@ -0,0 +1,2 @@
+mail.subject=password reset request
+mail.content=Dear :name \n You recently requested a password reset for your OpenMRS account. To complete the process, click the link below.\n  Please disregard this email if you didn't innitiate the password reset process.  \n
```
Comment (wluyima) [inline api/src/main/resources/emailTemplate.properties]: Change **Dear** to **Hi**

Is name a place holder? I would rather have it as **{name}** rather than **:name**

File: api/src/main/java/org/openmrs/api/UserService.java [hunk 24]
```diff
@@ -571,7 +572,7 @@
 	 * Sets a user's activation key
 	 * @param User The user for which the activation key will be set
 	 */
-	public User setUserActivationKey(User user);
+	public User setUserActivationKey(User user) throws MessageException;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/UserService.java]: Don't catch the exception

File: api/src/main/resources/emailTemplate.properties [hunk 25]
```diff
@@ -0,0 +1,2 @@
+mail.subject=password reset request
```
Comment (wluyima) [inline api/src/main/resources/emailTemplate.properties]: The file isn't supposed to be a properties file, otherwise you might as well move everything to the messages.properties file which already supports internationalization anyways and get rid of this file.

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 26]
```diff
@@ -1390,12 +1392,14 @@ public void getUserByUsernameOrEmail_shouldFailIfEmailIsNull() {
...
 	@Test
-	public void setUserActivationKey_shouldCreateUserActivationKey() {
+	public void setUserActivationKey_shouldCreateUserActivationKey() throws MessageException {
 		User u = new User();
 		u.setPerson(new Person());
 		u.addName(new PersonName("Benjamin", "A", "Wolfe"));
 		u.setUsername("bwolfe");
 		u.getPerson().setGender("M");
+		Context.getAdministrationService().setGlobalProperty(OpenmrsConstants.GP_HOST_URL,
+		    "http://localhost:8080/openmrs/admin/users/changePassword.form?activationkey");
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: I thought this is supposed to be {activationKey}

File: api/src/main/resources/emailTemplate.properties (hunk 23 above)
Comment (suthagar23) [reply api/src/main/resources/emailTemplate.properties]: @fanyui  The `\n` character is for the terminal outputs, will it work for the HTML based email content? Can you check and confirm with `<br/>`?

This link will active for some amount of time (24hrs). Right? So better to mention that at the end of this content.

Comment (wluyima) [reply api/src/main/resources/emailTemplate.properties]: That's what I think too, we should use html.

**Comment by wluyima** [comment]
@fanyui are you still working on this?

**Comment by fanyui** [comment]
@wluyima Yes am still working on it sorry for the absence, I had a fever(maleria) but have recovered. If am no longer working on a ticket i will identify.

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 27]
```diff
@@ -689,6 +694,9 @@ public static String getOpenmrsProperty(String property) {
 		props.add(new GlobalProperty(GP_MAIL_SMTP_STARTTLS_ENABLE, "false",
 		        "Set to true to enable TLS encryption, else set to false"));
 		
+		props.add(new GlobalProperty(GP_HOST_URL, "",
+		        "The URL to redirect to after requesting for a passwordreset. Always provide a place holder in this url with name {activationKey} so that it will be substituted by the actual activation key."));
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: The grammar is a little off around  **so that it will be substituted**,  how about **it will get substituted**?

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 28]
```diff
@@ -689,6 +694,9 @@ public static String getOpenmrsProperty(String property) {
 		props.add(new GlobalProperty(GP_MAIL_SMTP_STARTTLS_ENABLE, "false",
 		        "Set to true to enable TLS encryption, else set to false"));
 		
+		props.add(new GlobalProperty(GP_HOST_URL, "",
+		        "The URL to redirect to after requesting for a passwordreset. Always provide a place holder in this url with name {activationKey} so that it will get substituted by the actual activation key."));
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: I added a comment asking to rephrase this

File: api/src/main/resources/messages.properties [hunk 29]
```diff
@@ -3299,3 +3299,6 @@ Module.error.cyclicDependencies=Could not start some modules because they depend
 
 Condition.conditionShouldNotBeNull=Condition is required
 Condition.clinicalStatusShouldNotBeNull=Clinical status is required
+
+mail.passwordreset.subject=Password reset request
```
Comment (wluyima) [inline api/src/main/resources/messages.properties]: Password Reset should be enough

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java (hunk 28 above)
Comment (wluyima) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Did you see my comment about rephrasing this?

**Comment by fanyui** [comment]
@wluyima I have fixed this could you take a look?

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 30]
```diff
@@ -689,6 +694,9 @@ public static String getOpenmrsProperty(String property) {
 		props.add(new GlobalProperty(GP_MAIL_SMTP_STARTTLS_ENABLE, "false",
 		        "Set to true to enable TLS encryption, else set to false"));
 		
+		props.add(new GlobalProperty(GP_HOST_URL, "",
+		        "The URL to redirect to after requesting for a password reset. Always provide a place holder in this url with name {activationKey} it will get substituted by the actual activation key."));
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: You are missing a comma after {activationKey}

**Comment by wluyima** [comment]
Thanks @fanyui !

**Comment by fanyui** [comment]
@wluyima Am grateful for all the support, and eye opening  all-through.

**Comment by wluyima** [comment]
👍
